File: app/src/main/java/com/adisa/diningplus/adapters/FollowDialogAdapter.java
package com.adisa.diningplus.adapters;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ImageButton;
import android.widget.TextView;

import com.adisa.diningplus.R;

import java.util.ArrayList;
import java.util.HashSet;

/**
 * Created by Adisa on 5/4/2017.
 */

public class FollowDialogAdapter extends BaseAdapter {
    private Context context;
    private ArrayList<String> stringList = new ArrayList<>();
    private SharedPreferences preferences;

    private class ViewHolder {
        TextView name;
        ImageButton button;
    }

    public FollowDialogAdapter(Context context, HashSet<String> stringSet) {
        this.context = context;
        for (String string : stringSet) {
            stringList.add(string);
        }
        preferences = PreferenceManager.getDefaultSharedPreferences(context);
    }

    @Override
    public void notifyDataSetChanged() {
        super.notifyDataSetChanged();
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        LayoutInflater inflater = (LayoutInflater) context
                .getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        ViewHolder viewHolder;
        final String item = getItem(position);
        if (convertView == null) {
            viewHolder = new ViewHolder();
            convertView = inflater.inflate(R.layout.follow_dialog_item, null);
            viewHolder.name = (TextView) convertView.findViewById(R.id.followItemName);
            viewHolder.button = (ImageButton) convertView.findViewById(R.id.followItemRemove);
            convertView.setTag(viewHolder);
        } else {
            viewHolder = (ViewHolder) convertView.getTag();
        }
        viewHolder.name.setText(item);
        viewHolder.button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                HashSet<String> currentSet = (HashSet<String>) preferences.getStringSet("followedItems", new HashSet<String>());
                HashSet<String> newSet = new HashSet<String>();
                newSet.addAll(currentSet);
                newSet.remove(item);
                stringList.remove(item);
                SharedPreferences.Editor editor = preferences.edit();
                editor.putStringSet("followedItems", newSet);
                editor.apply();
                notifyDataSetChanged();
            }
        });

        return convertView;
    }

    @Override
    public int getCount() {
        return stringList.size();
    }

    @Override
    public String getItem(int position) {
        return stringList.get(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }
}

File: app/src/main/java/com/adisa/diningplus/adapters/FollowDialogAdapter.kt
package com.adisa.diningplus.adapters

import android.content.Context
import android.content.SharedPreferences
import android.preference.PreferenceManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.ImageButton
import android.widget.TextView
import com.adisa.diningplus.R
import java.util.*

class FollowDialogAdapter(private val context: Context, itemNameSet: MutableSet<String>) : BaseAdapter() {
    private val itemNames = ArrayList<String>()
    private val preferences: SharedPreferences

    private inner class ViewHolder {
        var name: TextView? = null
        var button: ImageButton? = null
    }

    override fun notifyDataSetChanged() {
        super.notifyDataSetChanged()
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup?): View {
        var convertView = convertView
        val inflater = context
                .getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        val viewHolder: ViewHolder
        val item = getItem(position)
        if (convertView == null) {
            viewHolder = ViewHolder()
            convertView = inflater.inflate(R.layout.follow_dialog_item, null)
            viewHolder.name = convertView.findViewById<View>(R.id.followItemName) as TextView
            viewHolder.button = convertView.findViewById<View>(R.id.followItemRemove) as ImageButton
            convertView.tag = viewHolder
        } else {
            viewHolder = convertView.tag as ViewHolder
        }
        viewHolder.name!!.text = item
        viewHolder.button!!.setOnClickListener {
            val currentSet = preferences.getStringSet("followed_items", HashSet()) as HashSet<String>?
            val newSet = HashSet<String>()
            newSet.addAll(currentSet!!)
            newSet.remove(item)
            itemNames.remove(item)
            val editor = preferences.edit()
            editor.putStringSet("followed_items", newSet)
            editor.apply()
            notifyDataSetChanged()
        }
        return convertView!!
    }

    override fun getCount(): Int {
        return itemNames.size
    }

    override fun getItem(position: Int): String {
        return itemNames[position]
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    init {
        itemNames.addAll(itemNameSet);
        preferences = PreferenceManager.getDefaultSharedPreferences(context)
    }
}